src/irr_ptr.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+#include <type_traits>
+#include "irrlichttypes.h"
+#include "IReferenceCounted.h"
+
+enum class Grab
+{
+	already_owned = 0,
+	do_grab = 1,
+};
+
+template <class ReferenceCounted, class = typename std::enable_if <
+	std::is_base_of <IReferenceCounted, ReferenceCounted>::value>::type>
+class irr_ptr
+{
+	ReferenceCounted *value = nullptr;
+public:
+	irr_ptr() = default;
+
+	irr_ptr(std::nullptr_t) noexcept
+	{
+	}
+
+	template <typename B, class = typename std::enable_if <
+		std::is_convertible <B *, ReferenceCounted *>::value>::type>
+	irr_ptr(const irr_ptr<B> &b) noexcept
+	{
+		reset(b.get());
+	}
+
+	template <typename B, class = typename std::enable_if <
+		std::is_convertible <B *, ReferenceCounted *>::value>::type>
+	irr_ptr(irr_ptr<B> &&b) noexcept
+	{
+		reset(b.release(), Grab::already_owned);
+	}
+
+	irr_ptr(ReferenceCounted *object, Grab grab) noexcept
+	{
+		reset(object, grab);
+	}
+
+	~irr_ptr()
+	{
+		reset();
+	}
+
+	template <typename B, class = typename std::enable_if <
+		std::is_convertible <B *, ReferenceCounted *>::value>::type>
+	irr_ptr &operator= (const irr_ptr<B> &b) noexcept
```

Comment:
Is that space after the `=` sign intentional?

Reply:
Yes. Does that look bad? Operator overloading is tricky to format well. `void operator()() {}` is perfectly valid.

Reply:
There is no official rule that I know of, but if you watch `operator=` as a function name, no spaces are required. `clang-format` wants to remove the space.